File: python/node.py
import numpy as np
import math


class Node:
    """
    Node class for Monte Carlo Tree Search (MCTS).

    Attributes:
        game (Game): The game object that provides game-specific logic.
        args (dict): A dictionary of arguments and hyperparameters for MCTS.
        state (np.ndarray): The current state of the game board.
        parent (Node): The parent node.
        action_taken (int): The action taken to reach this node.
        prior (float): The prior probability of selecting this node.
        children (list): The list of child nodes.
        visit_count (int): The number of times this node has been visited.
        value_sum (float): The sum of values backpropagated to this node.
    """
    def __init__(self, game, args, state, parent=None, action_taken=None, prior=0, visit_count=0):
        """
        Initializes the Node object with the given parameters.

        Args:
            game (Game): The game object.
            args (dict): A dictionary of arguments and hyperparameters.
            state (np.ndarray): The current state of the game board.
            parent (Node, optional): The parent node. Defaults to None.
            action_taken (int, optional): The action taken to reach this node. Defaults to None.
            prior (float, optional): The prior probability of selecting this node. Defaults to 0.
            visit_count (int, optional): The number of times this node has been visited. Defaults to 0.
        """
        self.game = game
        self.args = args
        self.state = state
        self.parent = parent
        self.action_taken = action_taken
        self.prior = prior
        self.children = []

        self.visit_count = visit_count
        self.value_sum = 0

    def is_expanded(self):
        """
        Checks if the node has been expanded (i.e., has child nodes).

        Returns:
            bool: True if the node has child nodes, False otherwise.
        """
        return len(self.children) > 0

    def select(self):
        """
        Selects the child node with the highest Upper Confidence Bound (UCB).

        Returns:
            Node: The child node with the highest UCB.
        """
        best_child = None
        best_ucb = -np.inf

        for child in self.children:
            ucb = self.get_ucb(child)
            if ucb > best_ucb:
                best_child = child
                best_ucb = ucb

        return best_child

    def get_ucb(self, child):
        """
        Calculates the Upper Confidence Bound (UCB) for a child node.

        Args:
            child (Node): The child node.

        Returns:
            float: The UCB value for the child node.
        """
        if child.visit_count == 0:
            q_value = 0
        else:
            q_value = 1 - ((child.value_sum / child.visit_count) + 1) / 2
        return q_value + self.args['C'] * (math.sqrt(self.visit_count) / (child.visit_count + 1)) * child.prior

    def expand(self, policy):
        """
        Expands the node by creating child nodes based on the given policy.
        """
        for action, prob in enumerate(policy):
            if prob > 0:
                child_state = self.state.copy()
                child_state = self.game.get_next_state(child_state, action, 1)
                child_state = self.game.change_perspective(child_state, player=-1)

                child = Node(self.game, self.args, child_state, self, action, prob)
                self.children.append(child)

    def backpropagate(self, value):
        """
        Backpropagates the value through the tree, updating visit counts and value sums.

        Args:
            value (float): The value to backpropagate.
        """
        self.value_sum += value
        self.visit_count += 1

        if self.parent is not None:
            value = self.game.get_opponent_value(value)
            self.parent.backpropagate(value)

File: Node.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * The Node class represents a node in the search tree of MCTS algorithm.
 */
public class Node {
    // Reference to the parent node
    private final Node parent;
    // Current state of the game
    private final State state;
    // number of times the node has been visited
    private int visitCount;
    // Number of wins for the red player from the current node
    private int sumWinsRed;
    // Number of wins for the blue player from the current node
    private int sumWinsBlue;
    // list of child nodes of the current node
    private final List<Node> childNodes;
    private List<Integer> possibleActions;
    private final int action; // the action that led to this node

    /**
     * Constructor for the root node.
     */
    public Node(State state) {
        this.parent = null;
        this.state = state;
        this.sumWinsRed = 0;
        this.sumWinsBlue = 0;
        this.childNodes = new ArrayList<>();
        this.action = -1;
    }

    /**
     * Constructor for a child node.
     */
    public Node(Node parentNode, State state, int action) {
        this.parent = parentNode;
        this.state = state;
        this.sumWinsRed = 0;
        this.sumWinsBlue = 0;
        this.childNodes = new ArrayList<>();
        this.action = action;
    }


    /**
     * This method checks if the node is fully expanded, i.e., if all its children have been visited.
     *
     * @return true if the node is fully expanded, false otherwise
     */
    public boolean isFullyExpanded() {
        return this.childNodes.size() == this.state.getPossibleActions().size();
    }

    /**
     * This method checks if the node is terminal, i.e., if the game has ended.
     *
     * @return true if the node is terminal, false otherwise
     */
    public boolean isTerminal() {
        return state.isTerminal();
    }

    /**
     * This method selects the child node with the highest Upper Confidence Bound (UCB) value.
     *
     * @return the best child node
     */
    public Node select() {
        return this.childNodes.parallelStream().max((child1, child2) -> {
            double ucb1 = child1.calculateUCB();
            double ucb2 = child2.calculateUCB();
            return Double.compare(ucb1, ucb2);
        }).orElse(null);
    }

    /**
     * This method calculates the Upper Confidence Bound (UCB) value of a given child node.
     *
     * @return the UCB value of the child node
     */
    public double calculateUCB() {
        double winRate = parent.state.getCurrentPlayer() ? ((double) sumWinsRed) / visitCount : ((double) sumWinsBlue) / visitCount;
        double exploration = Math.sqrt(Math.log(parent.visitCount) / visitCount);
        return winRate + Arguments.C * exploration;
    }

    /**
     * This method expands the current node by adding a new child node for a random action that has not been explored yet.
     */
    public Node expand() {
        if (possibleActions == null) {
            possibleActions = state.getPossibleActions();
        }
        int action = possibleActions.remove(new Random().nextInt(possibleActions.size()));
        State nextState = state.applyAction(action);
        Node childNode = new Node(this, nextState, action);
        this.childNodes.add(childNode);
        return childNode;
    }


    /**
     * This method selects a random child node of the current node.
     * This Method is used in the simulation phase of the MCTS algorithm.
     * @return the randomly selected child node
     */
    public Node getRandomChild() {
        List<Integer> actions = state.getPossibleActions();
        int action = actions.get(new Random().nextInt(actions.size()));
        State nextState = state.applyAction(action);
        return new Node(this, nextState, action);
    }

    /**
     * Returns the action that led to this node.
     *
     * @return The action that led to this node.
     */
    public int getAction() {
        return action;
    }

    /**
     * Returns the result of the game state at this node.
     *
     * @return The result of the game state at this node.
     */
    public boolean getResult() {
        return state.getResult();
    }

    /**
     * Updates the statistics of this node based on the result of a simulation.
     *
     * @param isRedWin A boolean indicating whether the red player won in the simulation.
     */
    public void updateStats(boolean isRedWin) {
        if (isRedWin) {
            sumWinsRed++;
        } else {
            sumWinsBlue++;
        }
        visitCount++;
    }

    /**
     * Returns the parent of this node.
     *
     * @return The parent of this node.
     */
    public Node getParent() {
        return parent;
    }

    /**
     * Returns the child nodes of this node.
     *
     * @return The child nodes of this node.
     */
    public List<Node> getChildNodes() {
        return childNodes;
    }

    /**
     * Returns the number of times this node has been visited.
     *
     * @return The number of times this node has been visited.
     */
    public int getVisitCount() {
        return visitCount;
    }

}
